### DISCUSSION TITLE
AbortController and AbortSignal globals

### DISCUSSION THREAD

**Discussion by TaylorBeeston** [MetaMask/snaps#678] [Ideas]
Hi there 👋 I'm Taylor from the LEF, working on making a [LearnCard](https://www.npmjs.com/package/@learncard/core) snap!

While building out the snap, I managed to work around every issue I came across except one: `AbortController` and `AbortSignal` not being globally available. `@learncard/core` internally depends on Ceramic libraries, which internally have a wrapper function around `fetch` that automatically times out via the use of an `AbortController`. In the past I've managed to get around environments that don't have the `AbortController` global available by using the `abort-controller` package to polyfill it, but when doing this inside of the SES runtime, I am running into this error when the fetch call gets made:

![Abort Signal Error](https://user-images.githubusercontent.com/39720479/184017676-1c4c3a14-f7b1-4bb5-86a4-c6bfa7424c70.png)

I dug around _a lot_ to try and find where this error is thrown, and I eventually found myself in the Chromium source code, where it _seems_ like the native `fetch` object will _only_ want to accept native `AbortSignal` objects (not "fake" `AbortSignal` objects created by a polyfill library with JS 😥). I'm not very good at reading C++ code (particularly a codebase as complicated as that!), so I could be wrong in that assumption, but that is where I landed!

So, for now, I am using `sed` against my bundled snap file to actually remove that `AbortSignal` object from being passed in when calling `fetch`, and everything is happy! This is, however, a pretty bad hack, and I would love it if we could instead just expose the browser's native `AbortController` and `AbortSignal` objects inside of the snap's SES runtime, as I believe that will fix the problem!

AFAIK, there's really no harm that one could cause if they got access to these two globals, though I am also definitely _not_ a security expert, and I am aware of how seemingly innocuous things can become malicious quickly with JS, so I understand that there would probably need to be some research into what possible security risks could be associated with adding these in.

tl;dr My snap is in a hacky, potentially buggy state until `AbortController` and `AbortSignal` can become globally available inside the snap runtime!

**Comment by ritave** [comment]
Hey @TaylorBeeston, we already had `AbortController` exposed, but we didn't expose `AbortSignal`.

I've added it now in #682, and so the next release of Flask should fix your errors.
